src/videojs-contrib-hls.js

```diff
@@ -117,110 +117,107 @@ const handleHlsLoadedMetadata = function(qualityLevels, hls) {
-    if (!variant.attributes ||
-        !variant.attributes.RESOLUTION ||
-        !variant.attributes.RESOLUTION.width ||
-        !variant.attributes.RESOLUTION.height) {
-      continue;
-    }
-
-    // since the playlists are sorted, the first variant that has
-    // dimensions less than or equal to the player size is the best
-    let variantResolution = variant.attributes.RESOLUTION;
+  // filter out playlists without resolution information
+  haveResolution = bandwidthPlaylists.filter(function(elem) {
+    return elem.attributes &&
+           elem.attributes.RESOLUTION &&
+           elem.attributes.RESOLUTION.width &&
+           elem.attributes.RESOLUTION.height;
+  });
 
-    if (variantResolution.width === width &&
-        variantResolution.height === height) {
-      // if we have the exact resolution as the player use it
-      resolutionPlusOne = null;
-      resolutionBestVariant = variant;
-      break;
-    } else if (variantResolution.width < width &&
-               variantResolution.height < height) {
-      // if both dimensions are less than the player use the
-      // previous (next-largest) variant
-      break;
-    } else if (!resolutionPlusOne ||
-               (variantResolution.width < resolutionPlusOneAttribute.width &&
-                variantResolution.height < resolutionPlusOneAttribute.height)) {
-      // If we still haven't found a good match keep a
-      // reference to the previous variant for the next loop
-      // iteration
-
-      // By only saving variants if they are smaller than the
-      // previously saved variant, we ensure that we also pick
-      // the highest bandwidth variant that is just-larger-than
-      // the video player
-      resolutionPlusOne = variant;
-      resolutionPlusOneAttribute = resolutionPlusOne.attributes.RESOLUTION;
-    }
+  // if we have the exact resolution as the player use it
+  resolutionBestVariantList = haveResolution.filter(function(elem) {
+    return elem.attributes.RESOLUTION.width === width &&
+           elem.attributes.RESOLUTION.height === height;
+  });
+  // ensure that we pick the highest bandwidth variant that have exact resolution
+  resolutionBestVariant = resolutionBestVariantList.filter(function(elem) {
+    return elem.attributes.BANDWIDTH === resolutionBestVariantList.slice(-1)[0].attributes.BANDWIDTH;
+  })[0];
+
+  // find the smallest variant that is larger than the player
+  // if there is no match of exact resolution
+  if (!resolutionBestVariant) {
+    resolutionPlusOneList = haveResolution.filter(function(elem) {
+      return elem.attributes.RESOLUTION.width > width ||
+             elem.attributes.RESOLUTION.height > height;
+    });
+    // find all the variants have the same smallest resolution
+    resolutionPlusOneSmallest = resolutionPlusOneList.filter(function(elem) {
+      return elem.attributes.RESOLUTION.width === resolutionPlusOneList[0].attributes.RESOLUTION.width &&
```

Comment:
While I think in most cases, `resolutionPlusOneList[0]` will be the smallest resolution, the original list of playlists is only sorted by bandwidth. This leaves the possibility that a variant with lower bandwidth but higher resolution would come first in the list.